onnx/defs/generator/defs.cc

```diff
@@ -92,26 +92,31 @@ ONNX_OPERATOR_SET_SCHEMA(
           const TensorProto* targetShapeInitializer = ctx.getInputData(0);
           if (!targetShapeInitializer) {
             // This is the case when exact shape input is not available.
-            // In this case, if the number of dimensions can be infered 
+            // In this case, if the number of dimensions can be infered
             // from the input 'shape' tensor, then we add the same number
             // of dimensions (without any dim_value information) to the output.
+            // (Rank inference only)
             if (ctx.getInputType(0)->tensor_type().has_shape()) {
-                auto& input_shape = getInputShape(ctx, 0);
-                auto input_shape_dim_size = input_shape.dim_size();
-                if(input_shape_dim_size > 1) {
-                    fail_shape_inference("Shape input must be a one-dimensional tensor.");
-                }
-                if (input_shape.dim(0).has_dim_value()) {
-                    const auto& input_shape_dim_value = input_shape.dim(0).dim_value();
-                    if (input_shape_dim_value > 0) {
-                      auto final_output_shape = 
-                          ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape();
-                        for (int i = 0; i < input_shape_dim_value; ++i) {
-                            auto newdim = final_output_shape->add_dim();
-                            (void)(newdim); // To eliminate "unused variable" compiler warning.
-                        }
-                    }
+              auto& input_shape = getInputShape(ctx, 0);
+              auto input_shape_dim_size = input_shape.dim_size();
+              if (input_shape_dim_size > 1) {
+                fail_shape_inference(
+                    "Shape input must be a one-dimensional tensor.");
+              }
+              if (input_shape.dim(0).has_dim_value()) {
+                const auto& input_shape_dim_value =
+                    input_shape.dim(0).dim_value();
+                if (input_shape_dim_value > 0) {
```

Comment:
May be worth adding a utility function "SetOutputRank" for this, since it will likely show up in many functions.